db/db_test.go

```diff
@@ -564,10 +578,104 @@ var _ = Describe("Database operation test", func() {
+					Expect(db.IsTemporaryDatabaseError(err)).To(BeTrue())
+					return err
+				})
+			Expect(err).NotTo(HaveOccurred())
+				return nil
+			})).To(Succeed())
+		})
+	})
+
+	Context("MySQL transaction retries", func() {
+		var (
+			firstConn  db.DB
+			secondConn db.DB
+
+			connOpts = options.Options{
+				RetryTxCount:    5,
+				RetryTxInterval: 0,
+			}
+		)
+		const (
+			networkTestID = "uniq-network-id"
+			serverTestID = "uniq-server-id"
+		)
+
+		JustBeforeEach(func() {
+			firstConn, err = dbutil.ConnectDB(mysqlType, mysqlConn, db.DefaultMaxOpenConn, connOpts)
+			Expect(err).ToNot(HaveOccurred())
+			secondConn, err = dbutil.ConnectDB(mysqlType, mysqlConn, db.DefaultMaxOpenConn, connOpts)
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, s := range manager.OrderedSchemas() {
+				Expect(firstConn.RegisterTable(s, false, true)).To(Succeed())
+			}
+			network1 := map[string]interface{}{
+				"id":                networkTestID,
+				"description":       "",
+				"shared":            false,
+				"route_targets":     []string{"1000:10000", "2000:20000"},
+				"providor_networks": map[string]interface{}{"segmentation_id": 10, "segmentation_type": "vlan"}}
+			nr, err := manager.LoadResource("network", network1)
+			Expect(err).ToNot(HaveOccurred())
+
+
+			server := map[string]interface{}{
+				"id":          serverTestID,
+				"network_id":  networkTestID,
+				"description": "",
+			}
+			sr, err := manager.LoadResource("server", server)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(db.WithinTx(firstConn, func(tx transaction.Transaction) error {
+				create(tx, nr)
+				create(tx, sr)
+				return nil
+			})).To(Succeed())
+		})
+
+		AfterEach(func() {
+			Expect(db.WithinTx(firstConn, func(tx transaction.Transaction) error {
+				Expect(tx.Delete(ctx, serverSchema, serverTestID)).To(Succeed())
+				Expect(tx.Delete(ctx, networkSchema, networkTestID)).To(Succeed())
+				return nil
+			})).To(Succeed())
+			firstConn.Close()
+			secondConn.Close()
+			schema.ClearManager()
+		})
+
+		BeforeEach(func() {
+			if os.Getenv("MYSQL_TEST") != "true" {
+				Skip("Test possible only on MySQL DB")
+			}
+			initSchemas()
+		})
+
+		It("should retry a few times after a deadlock", func() {
+			deadlockCount := 0
+			Expect(db.WithinTx(firstConn, func(firstTx transaction.Transaction) error {
+				err := db.WithinTx(secondConn, func(secondTx transaction.Transaction) error {
+					_, num, err := firstTx.LockList(ctx, networkSchema, nil, nil, nil, schema.SkipRelatedResources)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(num).ToNot(BeZero())
+
+					_, num, err = secondTx.LockList(ctx, serverSchema, nil, nil, nil, schema.SkipRelatedResources)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(num).ToNot(BeZero())
+
+					_, num, err = secondTx.LockList(ctx, networkSchema, nil, nil, nil, schema.SkipRelatedResources)
+					Expect(err).To(HaveOccurred())
+					deadlockCount++
+					if deadlockCount == 4 {
```

Comment:
`connOpts.RetryTxCount` instead of `4`. Unless you picked 4 so that it takes less time overall, in that case I think 2 or 3 would be just as fine.